### PR TITLE
Feature/aggregations2

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

This PR adds the possibility to perform custom feature aggregations in surrogates and use them for optimization.

An example application could be ContinuousDescriptorFeature or ContrinuousMolecularFeatures and the featurization of mixtures as discussed in Toby's recent paper: https://arxiv.org/pdf/2506.07619 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/experimental-design/bofire/blob/main/CONTRIBUTING.md#pull-requests)?

Y.

## Test Plan

Unit tests.

### CODE REVIEW

**Comment by jduerholt** [comment]
Note to myself: fix it for the multitask model and then also fit the standardization ...

File: bofire/data_models/domain/features.py [hunk 1]
```diff
@@ -241,6 +252,40 @@ class Features(_BaseFeatures[AnyFeature]):
...
 
+class EngineeredFeatures(_BaseFeatures[AnyEngineeredFeature]):
+    type: Literal["EngineeredFeatures"] = "EngineeredFeatures"  # type: ignore
+
+    def get_features2idx(self, offset: int = 0):
+        features2idx = {}
+        counter = offset
+        for feat in self.get():
+            features2idx[feat.key] = tuple(
+                (np.array(range(feat.n_outputs)) + counter).tolist(),
```
Comment (TobyBoyne) [inline bofire/data_models/domain/features.py]: I think this could be neater as `tuple(out_idx + counter for out_idx in range(feat.n_outputs))`

File: bofire/data_models/features/engineered_feature.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+from bofire.data_models.features.api import ContinuousDescriptorInput, ContinuousInput
+from bofire.data_models.features.feature import Feature
+from bofire.data_models.types import FeatureKeys
+
+
+class EngineeredFeature(Feature):
+    features: FeatureKeys
+    keep_features: bool = True
+
+    def validate_features(self, inputs: "Inputs"):  # noqa: F821 # type: ignore
```
Comment (TobyBoyne) [inline bofire/data_models/features/engineered_feature.py]: Can you not import `Inputs` wrapped in a `if TYPE_CHECKING` block to avoid needing these comments?

File: bofire/data_models/features/engineered_feature.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                f"The following features are missing in inputs: {missing_features}"
+            )
+        self._validate_features(inputs)
+
+    def _validate_features(self, inputs: "Inputs"):  # noqa: F821 # type: ignore
+        pass
+
+    @property
+    @abstractmethod
+    def n_outputs(self) -> int:
```
Comment (TobyBoyne) [inline bofire/data_models/features/engineered_feature.py]: I'm not sure I like `n_outputs` as a name - these aren't really outputs, just transformed inputs, right?

File: bofire/data_models/features/engineered_feature.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    order_id: ClassVar[int] = 1
+
+    @property
+    def n_outputs(self) -> int:
+        return 1
+
+
+class WeightedSumFeature(EngineeredFeature):
+    type: Literal["WeightedSumFeature"] = "WeightedSumFeature"
+    descriptors: Annotated[List[str], Field(min_length=2)]
```
Comment (TobyBoyne) [inline bofire/data_models/features/engineered_feature.py]: Can you use `data_models.types.Descriptors` here?

File: bofire/data_models/features/engineered_feature.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    def validate_features(self, inputs: "Inputs"):  # noqa: F821 # type: ignore
+        super().validate_features(inputs)
+        for feature_key in self.features:
+            feature = inputs.get_by_key(feature_key)
+            if not isinstance(feature, ContinuousDescriptorInput):
+                raise ValueError(
+                    f"Feature '{feature_key}' is not a ContinuousDescriptorInput",
+                )
+            if len(set(self.descriptors) - set(feature.descriptors)) > 0:  # type: ignore
```
Comment (TobyBoyne) [inline bofire/data_models/features/engineered_feature.py]: Should this not be `set(self.descriptors) == set(feature.descriptors)`? Is it okay if there are some descriptors present in the EngineeredFeature, which are not present in the Feature?

File: bofire/surrogates/engineered_features.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+from bofire.data_models.features.api import (
+    EngineeredFeature,
+    MeanFeature,
+    SumFeature,
+    WeightedSumFeature,
+)
+from bofire.data_models.types import InputTransformSpecs
+
+
+def map_sum_feature(
```
Comment (TobyBoyne) [inline bofire/surrogates/engineered_features.py:14]: Could these three mapping functions be rewritten as a single function? The only thing that changes between them is the definition of `f`.

File: bofire/surrogates/utils.py [hunk 7]
```diff
@@ -107,86 +108,145 @@ def get_categorical_feature_keys(
...
-                tf1=categorical_transform, tf2=scaler
-            )
-        elif categorical_transform is not None:
-            input_transform = categorical_transform
-        elif scaler is not None:
-            input_transform = scaler
-    else:
-        input_transform = scaler
-    return input_transform
+        if feature.keep_features is False:
```
Comment (TobyBoyne) [inline bofire/surrogates/utils.py]: `if not feature.keep_features`

File: tests/bofire/surrogates/test_engineered_features.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        inputs=inputs, transform_specs={}, feature=aggregation
+    )
+
+    orig = torch.tensor([[0.1, 0.2, 0.7], [0.4, 0.1, 0.5]]).to(**tkwargs)
+    result = aggregator(orig)
+
+    assert result.shape[0] == 2
+    assert result.shape[1] == 5
+
+    assert torch.allclose(result[:, :-2], orig)
```
Comment (TobyBoyne) [inline tests/bofire/surrogates/test_engineered_features.py]: Please add the expected return values

**Review by TobyBoyne — COMMENTED** [review]
Looks good, most of my comments are pretty minor nitpicks.

Since you mentioned the Catechol paper, I'm not sure this PR has enough features to replicate our model. We have two categorical inputs (solvent A and solvent B) with descriptors, then used a weighted sum of those two solvents. However, it looks to me like this PR only supports ContinuousDescriptors, which have fixed descriptors. Have I understood this correctly, and will you be adding this feature?

File: bofire/data_models/features/engineered_feature.py (hunk 2 above)
Comment (jduerholt) [reply bofire/data_models/features/engineered_feature.py]: Hmm, problem are here that I get circular imports ..., do you think this would help?

File: bofire/data_models/domain/features.py (hunk 1 above)
Comment (jduerholt) [reply bofire/data_models/domain/features.py]: I agree!

File: bofire/data_models/features/engineered_feature.py (hunk 2 above)
Comment (jduerholt) [reply bofire/data_models/features/engineered_feature.py]: It solved it, thanks for the hint!

File: bofire/data_models/features/engineered_feature.py (hunk 3 above)
Comment (jduerholt) [reply bofire/data_models/features/engineered_feature.py]: good point, I renamed it to `n_transformed_inputs` .

File: bofire/data_models/features/engineered_feature.py (hunk 4 above)
Comment (jduerholt) [reply bofire/data_models/features/engineered_feature.py]: Good catch!

File: bofire/surrogates/engineered_features.py (hunk 6 above)
Comment (jduerholt) [reply bofire/surrogates/engineered_features.py:14]: The `map_weighted_sum_feature` is different, for this reason, I would keep it as is.

File: tests/bofire/surrogates/test_engineered_features.py (hunk 8 above)
Comment (jduerholt) [reply tests/bofire/surrogates/test_engineered_features.py]: Done.

File: bofire/surrogates/utils.py (hunk 7 above)
Comment (jduerholt) [reply bofire/surrogates/utils.py]: Done.

File: bofire/data_models/features/engineered_feature.py (hunk 4 above)
Comment (jduerholt) [reply bofire/data_models/features/engineered_feature.py]: Done.

File: bofire/data_models/features/engineered_feature.py (hunk 3 above)
Comment (jduerholt) [reply bofire/data_models/features/engineered_feature.py]: Done.

**Comment by bertiqwerty** [comment]
I had a first look at your changes. I think, I need a little more context. With this PR I can use additional features computed from the input features for the surrogate model. These computations can only be 

- sum, 
- addition, 
- mean, and 
- weighted sum

correct? Now If I optimize the acquisition function, I optimize over the original inputs and have these as additional transforms. Did I get this right? And this only makes sense to be applied to continuous descriptors? Why?

**Comment by jduerholt** [comment]
@TobyBoyne:

 > I haven't been able to closely look at the changes to benchmark.py but it all seems okay from a brief glance.

This is also only slightly related, it gives some benchmarks to use the new machinery.

> You're right about the solvent mixtures. You also need to constrain the weighted sum to be == 1 (which I think you have done but I'm not sure), and in the case of only using two solvents must have an NChooseK constraint. 

Yeah, fully agree, but this is existing machinery. We have to work on improved handling of NChooseKs next. But I have an idea for this already.

> If I were able to do a full review now, I would double check that when you use the dimension-scaled prior, this doesn't scale when adding more ContinuousDescriptorInputs.

I am not sure, If I fully agree here, I give an few example on the current behavior:

If you have 10 solvents and each solvent is defined by three descriptors, and you add a WeightedSumFeature over all descriptors, you add three dimensions to the GP surrogate, in total you have then 13 dimensions in the GP, and the dimensionality scaled prior should also act on this 13. If you specify in the kernel of the surrogate, that only the three weighted descriptors should be considered, than the dimensionality scaled priors should act over dim 3.  Agree? This is how it currently works.

**Comment by jduerholt** [comment]
@bertiqwerty fine for you to go ahead with this? Or do you need more explanation?

**Review by bertiqwerty — APPROVED** [review]
Thanks Johannes.
